File: CS/WindowsFormsApplication2/CommentReader.cs
using DevExpress.Pdf;
using System.Collections.Generic;

namespace WindowsFormsApplication2 {
    public class CommentReader {
        public List<CommentData> ReadComments(string fileName) {
            List<CommentData> comments = new List<CommentData>();
            using (PdfDocumentProcessor pServer = new PdfDocumentProcessor()) {
                pServer.LoadDocument(fileName);
                for (int i = 0; i < pServer.Document.Pages.Count; i++) {
                    PdfPage page = pServer.Document.Pages[i];
                    int pageNumber = i + 1;
                    foreach (PdfAnnotation annotation in page.Annotations) {
                        if (annotation is PdfTextAnnotation) {
                            PdfTextAnnotation comment = (PdfTextAnnotation)annotation;
                            if (comment.InReplyTo == null) {
                                if (comments.FindIndex((item) => item.Name == comment.Name) == -1)
                                    comments.Add(CreateCommentData(comment, pageNumber));
                            }
                            else {
                                ReplyData replyData = CreateReplyData(comment, pageNumber);
                                PdfTextAnnotation parentAnnotation = (PdfTextAnnotation)comment.InReplyTo;
                                string name = parentAnnotation.Name;
                                CommentData parentComment = comments.Find((item) => item.Name == name);
                                if (parentComment == null) {
                                    parentComment = CreateCommentData(comment, pServer.Document.Pages.IndexOf(parentAnnotation.Page) + 1);
                                    comments.Add(parentComment);
                                }
                                parentComment.Replies.Add(replyData);
                            }
                        }
                    }
                }
            }
            return comments;
        }
        ReplyData CreateReplyData(PdfTextAnnotation comment, int pageNumber) {
            ReplyData reply = new ReplyData();
            InitData(reply, comment, pageNumber);
            return reply;
        }
        CommentData CreateCommentData(PdfTextAnnotation comment, int pageNumber) {
            CommentData data = new CommentData();
            InitData(data, comment, pageNumber);
            PdfPoint pageCropBoxBottomLeft = comment.Page.CropBox.BottomLeft;
            PdfRectangle rectangle = new PdfRectangle(comment.Rect.Left - pageCropBoxBottomLeft.X, comment.Rect.Bottom - pageCropBoxBottomLeft.Y, comment.Rect.Right - pageCropBoxBottomLeft.X, comment.Rect.Top - pageCropBoxBottomLeft.Y);
            data.BoundingBox = rectangle;
            return data;
        }
        void InitData(ReplyData data, PdfTextAnnotation comment, int pageNumber) {
            data.Subject = comment.Title;
            data.Text = comment.Contents;
            data.Name = comment.Name;
            data.CreationDate = comment.CreationDate;
            data.PageNumber = pageNumber;
        }

    }
}

File: VB/WindowsFormsApplication2/CommentReader.vb
Imports DevExpress.Pdf
Imports System.Collections.Generic

Namespace WindowsFormsApplication2
    Public Class CommentReader
        Public Function ReadComments(ByVal fileName As String) As List(Of CommentData)
            Dim comments As New List(Of CommentData)()
            Using pServer As New PdfDocumentProcessor()
                pServer.LoadDocument(fileName)
                For i As Integer = 0 To pServer.Document.Pages.Count - 1
                    Dim page As PdfPage = pServer.Document.Pages(i)
                    Dim pageNumber As Integer = i + 1
                    For Each annotation As PdfAnnotation In page.Annotations
                        If TypeOf annotation Is PdfTextAnnotation Then
                            Dim comment As PdfTextAnnotation = CType(annotation, PdfTextAnnotation)
                            If comment.InReplyTo Is Nothing Then
                                If comments.FindIndex(Function(item) item.Name = comment.Name) = -1 Then
                                    comments.Add(CreateCommentData(comment, pageNumber))
                                End If
                            Else
                                Dim replyData As ReplyData = CreateReplyData(comment, pageNumber)
                                Dim parentAnnotation As PdfTextAnnotation = CType(comment.InReplyTo, PdfTextAnnotation)
                                Dim name As String = parentAnnotation.Name
                                Dim parentComment As CommentData = comments.Find(Function(item) item.Name = name)
                                If parentComment Is Nothing Then
                                    parentComment = CreateCommentData(comment, pServer.Document.Pages.IndexOf(parentAnnotation.Page) + 1)
                                    comments.Add(parentComment)
                                End If
                                parentComment.Replies.Add(replyData)
                            End If
                        End If
                    Next annotation
                Next i
            End Using
            Return comments
        End Function
        Private Function CreateReplyData(ByVal comment As PdfTextAnnotation, ByVal pageNumber As Integer) As ReplyData
            Dim reply As New ReplyData()
            InitData(reply, comment, pageNumber)
            Return reply
        End Function
        Private Function CreateCommentData(ByVal comment As PdfTextAnnotation, ByVal pageNumber As Integer) As CommentData
            Dim data As New CommentData()
            InitData(data, comment, pageNumber)
            Dim pageCropBoxBottomLeft As PdfPoint = comment.Page.CropBox.BottomLeft
            Dim rectangle As New PdfRectangle(comment.Rect.Left - pageCropBoxBottomLeft.X, comment.Rect.Bottom - pageCropBoxBottomLeft.Y, comment.Rect.Right - pageCropBoxBottomLeft.X, comment.Rect.Top - pageCropBoxBottomLeft.Y)
            data.BoundingBox = rectangle
            Return data
        End Function
        Private Sub InitData(ByVal data As ReplyData, ByVal comment As PdfTextAnnotation, ByVal pageNumber As Integer)
            data.Subject = comment.Title
            data.Text = comment.Contents
            data.Name = comment.Name
            data.CreationDate = comment.CreationDate
            data.PageNumber = pageNumber
        End Sub

    End Class
End Namespace
